File: code/ex881.cpp
class Solution {
public:
    int numRescueBoats(vector<int>& people, int limit) {
        int n = people.size();
        sort(people.begin(), people.end());
        int cnt = 0, lo = 0, hi = n - 1;
        while (lo <= hi) {
            int l = limit - people[hi--];
            if (lo <= hi && l - people[lo] >= 0) {
                l -= people[lo++];
            }
            cnt++;
        }
        return cnt;
    }
};

File: code/ex881.java
class Solution {
    public int numRescueBoats(int[] people, int limit) {
        int n = people.length;
        Arrays.sort(people);
        int lo = 0, hi = n - 1, cnt = 0;
        while (lo < hi) {
            if (people[lo] + people[hi] <= limit) {
                lo++;    hi--;
            } else
                hi--;
            cnt++;
        }
        return lo == hi ? cnt + 1 : cnt;
    }
}